internal/socket.go

```diff
@@ -140,3 +94,84 @@ func StringToUcred(ucredStr string) (unix.Ucred, error) {
 
 	return unix.Ucred{Pid: int32(pid), Uid: uint32(uid), Gid: uint32(gid)}, nil
 }
+
+// getUnixCreds returns info from unix socket connection about the process on the other end.
+func getUnixCreds(conn net.Conn) (*unix.Ucred, error) {
+	conn2 := extractConnection(conn)
+	if conn2 == nil {
+		return nil, fmt.Errorf("cannot extract connection of proper type, is netutil.LimitListener used?")
+	}
+
+	unixConn, ok := conn2.(*net.UnixConn)
+	if !ok {
+		return nil, fmt.Errorf("socket is not a unix socket")
+	}
+
+	rawConn, err := unixConn.SyscallConn()
+	if err != nil {
+		return nil, fmt.Errorf("getting raw connection: %w", err)
+	}
+
+	var ucred *unix.Ucred
+	var internalErr error
+	err = rawConn.Control(func(fd uintptr) {
+		ucred, internalErr = unix.GetsockoptUcred(int(fd), unix.SOL_SOCKET, unix.SO_PEERCRED)
+	})
+	if internalErr != nil {
+		return nil, fmt.Errorf("doing GetsockoptUcred: %w", internalErr)
+	}
+	if err != nil {
+		return nil, fmt.Errorf("doing rawConn Control: %w", err)
+	}
+
+	if err := authenticateUser(ucred, []string{"nordvpn"}); err != nil {
+		return nil, err
+	}
+
+	return ucred, nil
+}
+
+func authenticateUser(ucred *unix.Ucred, allowGroups []string) error {
+	// root?
+	if ucred.Uid == 0 {
+		return nil
+	}
+	userInfo, err := user.LookupId(fmt.Sprintf("%d", int(ucred.Uid)))
+	if err != nil {
+		return fmt.Errorf("authenticate user, lookup user info: %s", err)
+	}
+	// user belongs to the allowed group?
+	groups, err := userInfo.GroupIds()
+	if err != nil {
+		return fmt.Errorf("authenticate user, check user groups: %s", err)
+	}
+	for _, groupId := range groups {
+		groupInfo, err := user.LookupGroupId(groupId)
+		if err != nil {
+			return fmt.Errorf("authenticate user, check user group: %s", err)
+		}
+		for _, allowGroupName := range allowGroups {
+			if groupInfo.Name == allowGroupName {
+				return nil
+			}
+		}
+	}
+	return fmt.Errorf("requesting user does not have permissions")
+}
+
+// using `netutil.LimitListener“ but it has internal unexported type `limitListenerConn`
+// which wraps original `net.UnixConn` value inside by embeding abstract interface `net.Conn`
+// this way we cannot access `net.UnixConn` value, because of that we use `go reflection`
+// to extract original wrapped value.
+func extractConnection(c interface{}) net.Conn {
```

Comment:
would it make sense to have a test for this, to catch if `LimitListener` changes?

Reply:
Good point, thx